### PR TITLE
flat prefetch in shell, flat imports in entrypoints

### LINKED ISSUE #72: Unbundled build: insert link headers
The `main` file should have `<link rel="prefetch"` tags added for each file in the shared bundle and main app files.

Each additional `bundle` should have `<link rel="import"` tags added for each file that they transitively depend on that isn't already prefetched in `main`

### PR DESCRIPTION
**Pull request by azakus**
Fixes #72

### CODE REVIEW

File: src/build/build.ts [hunk 1]
```diff
@@ -119,11 +120,26 @@ export function build(options?: BuildOptions): Promise<any> {
...
+      .pipe(prefetchTransform)
       .pipe(vfs.dest('build/unbundled'))
 
     let bundledPhase = new ForkedVinylStream(allFiles)
       .pipe(bundler.bundle)
+      .on('finish', () => {
+        bundler._entrypointToDeps.then((map) => {
+          // let shellDeps = map.get(shell).map((p) => path.relative(root, p));
+          // map.set(main, shellDeps);
+          map.set(main, map.get(shell));
```
Comment (justinfagnani) [inline src/build/build.ts]: add a comment explaining my main is added to this map

File: src/build/build.ts [hunk 2]
```diff
@@ -119,11 +120,26 @@ export function build(options?: BuildOptions): Promise<any> {
...
+    let prefetchTransform = new PrefetchTransform(root, [main], entrypoints, prefetchDepsPromise);
     let unbundledPhase = new ForkedVinylStream(allFiles)
+      .pipe(prefetchTransform)
       .pipe(vfs.dest('build/unbundled'))
 
     let bundledPhase = new ForkedVinylStream(allFiles)
       .pipe(bundler.bundle)
+      .on('finish', () => {
+        bundler._entrypointToDeps.then((map) => {
+          // let shellDeps = map.get(shell).map((p) => path.relative(root, p));
```
Comment (justinfagnani) [inline src/build/build.ts]: remove?

File: src/build/prefetch.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    this.importTargets = importTargets;
+    this.dependencyMapPromise = dependencyMapPromise;
+    this.fileMap = new Map<string, File>();
+  }
+
+  pullUpDeps(
+    file: File,
+    deps: string[],
+    type: 'import' | 'prefetch'
+  ){
```
Comment (justinfagnani) [inline src/build/prefetch.ts]: nit: add space

File: src/build/prefetch.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    let contents = file.contents.toString();
+    let ast = dom5.parse(contents);
+    const head = dom5.query(ast, dom5.predicates.hasTagName('head'));
```
Comment (justinfagnani) [inline src/build/prefetch.ts]: why `const`? lets just be consistent with locals

File: src/build/prefetch.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import File = require('vinyl');
+
+export class PrefetchTransform extends Transform {
+  root: string;
+  prefetchTargets: string[];
+  importTargets: string[];
+  fileMap: Map<string, File>;
+  dependencyMapPromise: Promise<Map<string, string[]>>;
+
+  constructor(
```
Comment (justinfagnani) [inline src/build/prefetch.ts]: comment the parameters

File: src/build/prefetch.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+  _flush(done: (err?) => void) {
+    if (this.fileMap.size === 0) {
+      return done();
+    }
+    this.dependencyMapPromise.then((map) => {
+      for (let prefetch of this.prefetchTargets) {
+        let file = this.fileMap.get(prefetch);
+        let deps = map.get(prefetch);
+        // console.log('prefetch', file.path, deps);
```
Comment (justinfagnani) [inline src/build/prefetch.ts]: remove

File: src/build/prefetch.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        if (deps) {
+          this.pullUpDeps(file, deps, 'prefetch');
+        }
+        this.push(file);
+        this.fileMap.delete(prefetch);
+      }
+      for (let im of this.importTargets) {
+        let file = this.fileMap.get(im);
+        let deps = map.get(im);
+        // console.log('import', file.path, deps);
```
Comment (justinfagnani) [inline src/build/prefetch.ts]: remove

File: src/build/prefetch.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        if (deps) {
+          this.pullUpDeps(file, deps, 'import');
+        }
+        this.push(file);
+        this.fileMap.delete(im);
+      }
+      for (let leftover of this.fileMap.keys()) {
+        console.log('unused', leftover);
```
Comment (justinfagnani) [inline src/build/prefetch.ts]: expand on the "unused" bit

**Comment by justinfagnani** [comment]
LGTM

File: src/build/prefetch.ts (hunk 5 above)
Comment (azakus) [inline src/build/prefetch.ts]: Done.

**Pull request merged**